feat(autoware_bevfusion): implementation of bevfusion using tensorrt

knzo25 opened the pull request:
## Description

This PR introduces BEVFusion to autoware using TensorRT.
I would like to ask reviewers to let the "integration" into the pipeline/launchers to a posterior PR :pray: 

## Related links

**Parent Issue:**

- Link

<!-- ⬇️🟢
**Private Links:**

- [CompanyName internal link]()
⬆️🟢 -->

## How was this PR tested?

## Notes for reviewers

The onnx files can be found here: https://drive.google.com/drive/folders/1p2wrAMSubDC-AgInTN7WfJuvp52LBiVq. 
Once the PR is approved but before merging it, I will add the final model like all the current ones.

Since this package introduces early fusion, it can not directly be integrated into autoware (the lidar-only model can). Such integration should be relegated to the next PR to avoid increasing unnecessarily the number of stakeholders on this PR.

To test the PR, I recommend using the taxi project (will omit the launch command), and launch bevfusion separatedly.
```bash
ros2 launch autoware_lidar_bevfusion lidar_bevfusion.launch.xml
```
For now, the models must be placed in the `config` folder, and to change the modality (default is camera-lidar), the `yaml' file can be modified. This is the yaml file parameters needed for the lidar-only model.

```yaml
/**:
  ros__parameters:
    # modality
    sensor_fusion: false
    # non-network params
    max_camera_lidar_delay: 0.12
    # plugins
    plugins_path: $(find-pkg-share autoware_lidar_bevfusion)/plugins/libautoware_tensorrt_plugins.so
    # network
    trt_precision: fp16
    cloud_capacity: 2000000
    onnx_path: "$(var model_path)/bevfusion_lidar_v2.onnx"
    engine_path: "$(var model_path)/bevfusion_lidar_v2.engine"
    # pre-process params
    densification_num_past_frames: 0
    densification_world_frame_id: map
    # post-process params
    circle_nms_dist_threshold: 0.5
    iou_nms_target_class_names: ["CAR"]
    iou_nms_search_distance_2d: 10.0
    iou_nms_threshold: 0.1
    yaw_norm_thresholds: [0.3, 0.3, 0.3, 0.3, 0.0] # refers to the class_names
    score_threshold: 0.1

```

## Interface changes

None.

<!-- ⬇️🔴

### Topic changes

#### Additions and removals

| Change type   | Topic Type      | Topic Name    | Message Type        | Description       |
|:--------------|:----------------|:--------------|:--------------------|:------------------|
| Added/Removed | Pub/Sub/Srv/Cli | `/topic_name` | `std_msgs/String`   | Topic description |

#### Modifications

| Version | Topic Type      | Topic Name        | Message Type        | Description       |
|:--------|:----------------|:------------------|:--------------------|:------------------|
| Old     | Pub/Sub/Srv/Cli | `/old_topic_name` | `sensor_msgs/Image` | Topic description |
| New     | Pub/Sub/Srv/Cli | `/new_topic_name` | `sensor_msgs/Image` | Topic description |

### ROS Parameter Changes

#### Additions and removals

| Change type   | Parameter Name | Type     | Default Value | Description       |
|:--------------|:---------------|:---------|:--------------|:------------------|
| Added/Removed | `param_name`   | `double` | `1.0`         | Param description |

#### Modifications

| Version | Parameter Name   | Type     | Default Value | Description       |
|:--------|:-----------------|:---------|:--------------|:------------------|
| Old     | `old_param_name` | `double` | `1.0`         | Param description |
| New     | `new_param_name` | `double` | `1.0`         | Param description |

🔴⬆️ -->

## Effects on system behavior

None.

xmfcx commented:
- Is this related to https://github.com/autowarefoundation/autoware.universe/pull/7956 ?

knzo25 commented:
As TensorRT was upgraded and spconv was added (https://github.com/autowarefoundation/autoware/pull/5794), I will be opening this PR :tada:

knzo25 commented:
@amadeuszsz 
As discussed internally, I pasted the links and how to execute the model. Please let me know if you need anything :pray:

freejumperd commented:
@knzo25 thank you for this great work. If I may ask a quick question. In terms of bevfusion inference through trt, what's the main difference between the version under your development:[https://github.com/knzo25/autoware.universe/tree/feat/bevfusion/perception/autoware_lidar_bevfusion](url)
And the official NVDA AI IOT one:
[https://github.com/NVIDIA-AI-IOT/Lidar_AI_Solution/tree/master/CUDA-BEVFusion/src/bevfusion](url) 
And I assume there is no model modification between autoware version and MIT original one? Please suggest. Thank you!

knzo25 commented:
@freejumperd 

The main difference with `NVIDA AI IOT`'s implementation is that they use a closed source shared library for inference (sparse convolutions), even if it is a public binary. Another PR with that implementation may be sent later by another group of contributors as far as I understand. A big difference in terms of dvelopment and deployment of models, is that this PR handles models generated by our ml stack (https://github.com/tier4/AWML/tree/main/projects/BEVFusion) directly.

With respect to the original implementation, those questions may be better suited for our ml stack rather than the inference node here. But in a few words, the lidar model more or less remains the same, although we have evaluated bigger models for offline purposes (we have also tweaked some minor things that have increased performance). The camera-lidar model was too focused on nuscenes, so there have been a few other developments as well, but nothing so big.

freejumperd commented:
@knzo25 thanks for active response knzo. So in short, just like the centerpoint inference node already published through autoware universe, the actualy inference cpp and cu code has been modifed and optimised by you and rest of community ? The nvda one may be suitable for exact original model trained on nuscense dataset but certainly not as efficient as for autoware retrained centerpoint or bevfusion should we say? And when would you expect this bevfusion trt inference + ros2 node to be available open source ?

knzo25 commented:
@freejumperd 
More information about the nvidia's implementation vs. this approach was presented in an a previous issue (do not have the link now though). I would separate nvidia's inference code with the original MIT implementation though. I just meant that the original config files and some of the processing in the camera-lidar pipeline do not work well with our vehicle's config.

As for when this will be available open source? In a way, it already is, since you can use this branch under the Apache license. As for when will this be merged, it is really up to the reviewers. If you want to play a part in that, we really would appreciate it!

freejumperd commented:
@knzo25 thanks for explaining. I am certainly interested and plan to go through the recently published AWML pipeline and then running inference test to check FPS etc.
One last thing if I may ask here (more relavent on model side then trt). I was playing with original MIT model+weights but with our own dataset and modified data process pipeline, without touching the model architecture (different number for cameras, different intrinsic +extrinsic 6DOF, different model of top lidar etc) the zero shot performance was extremely bad and almost like no detection. So I wonder for the AWML version trained with T4+nuscense dataset (e.g. both with same amount of 6 cameras + 1 top lidar), if again I test with our own dataset/pipeline (e.g. 7 cameras, 2 lidars etc) and using the available pth (without touching the model), would you expect the AWML bevfusion model zoo would provide good generallizaiton ability? Or if e.g. the sensor suites are different (number of camera/lidar, 6DOF etc), retrain the model with own dataset would be a must ? If so then the only best way if I want to use the open pth, is to make your vehicle sensor suites as close as possible to nuscense / T4 taxi ?

knzo25 commented:
@freejumperd 
(Current) sensor fusion models, especially early fusion ones, are extremely dependent on the vehicle configuration. That is especially true for BEVFusion since it unprojects  points from the camera to the BEV space. That being said, in my experience, for this method, the image features do not an impact so strong that there are no detections.
If I had to say, I would suspect an error on your pipeline (beware of intensity profiles and so on). Further support would require seeing the data and formal consulting hours (If our institutions have an agreement, we can continue this conversation on other channels).
Regarding whether retraining is a must, it really depends how you handle the data pipeline to reduce the mismatches (to some degree this can be done), but since we are not really training a foundational model here, I would always recommend fine tuning or retraining.

freejumperd commented:
@knzo25 thanks again, what's been discussed are really helpful! Let me leverage with AWML pipeline first to revise our own data pipeline. Will come back to you later for any potential collaboration on the perception topics.

knzo25 commented:
@freejumperd 
Ahh, I just forgot something about the open source's repository implementation of BEVFusion. While the repository was finally open-sourced, the final pipeline for the camera-lidar model and export logic have not been updated (was developing in a branch on the private repo). I should send a PR soon-ish.

freejumperd commented:
@knzo25 amazing! You guys really have done some great work for promoting such advanced early fusion architecture for open source community. Really looking forward to learn more out of it. And also looking for any coming good work leveraging with VLMs 😉

mitsudome-r commented:
@knzo25 Is it possible for you to make the onnx model public?
https://drive.google.com/drive/folders/1o7sxrIx7rILe00KuPr-odS_NNFYPj26p

Zhao-Qihao commented:
@knzo25 thanks for your great work! I want to know when the onnx model while be public?

knzo25 commented:
@mitsudome-r @Zhao-Qihao 
Sorry, I will try to upload a model today or tomorrow (In the awml repo just got a better model but had no time to onxx-fy it yet)

Zhao-Qihao commented:
> @mitsudome-r @Zhao-Qihao Sorry, I will try to upload a model today or tomorrow (In the awml repo just got a better model but had no time to onnx-fy it yet)

@knzo25 Appreciate! Another question is that the training code is in awml repo and the deploy in tensorrt code is in your repo?

knzo25 commented:
@Zhao-Qihao 
Training and onnx generation in awml - inference in this PR. However, the latest deployment (pth -> onnx) branch with some other improvements have not been merged yet (it is in an internal version of the repo before it being public so I need to cherrypick and adapt some parts. I am getting presure to do it, but there are several things moving at the moment :sweat_smile: )

Zhao-Qihao commented:
> @Zhao-Qihao Training and onnx generation in awml - inference in this PR. However, the latest deployment (pth -> onnx) branch with some other improvements have not been merged yet (it is in an internal version of the repo before it being public so I need to cherrypick and adapt some parts. I am getting presure to do it, but there are several things moving at the moment 😅 )

@knzo25 OK, thanks again

Zhao-Qihao commented:
@knzo25 Sorry to bother you, have you uploaded the onnx file? I want to test it in my local machine.

knzo25 commented:
@mitsudome-r 
Sorry, I though you had made a new public link. I will ask the IT dept to make it public.

knzo25 commented:
Updated the link in the description, but will also paste it here
https://drive.google.com/drive/folders/1p2wrAMSubDC-AgInTN7WfJuvp52LBiVq

However, this is not the latest model. Will update when possible

amadeuszsz commented on perception/autoware_lidar_bevfusion/src/bev_ops/bev_pool_cuda.cu:
```diff
@@ -0,0 +1,57 @@
+// Available under Apache-2.0 license
+
+#include <stdio.h>
+#include <stdlib.h>
```
In general, use c++ style headers instead. Actually, in this file both headers are unnecessary.

knzo25 replied:
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/97d33c3611b81fbe2042cb7f0f68b805b9d18c2b

amadeuszsz commented on perception/autoware_lidar_bevfusion/src/bev_ops/bev_pool_cuda.cu:
```diff
@@ -0,0 +1,57 @@
+  const int * geom_feats, const int * interval_starts, const int * interval_lengths, float * out,
+  cudaStream_t & stream)
+{
+  bev_pool_kernel<<<(int)ceil(((double)n_intervals * c / 256)), 256, 0, stream>>>(
```
Can you use your `divUp` function instead? If not, add `cmath` header.

knzo25 replied:
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/d7565bcc34165705f256ca947df42a8a8fad5368.

I wanted to avoid modifying the file, but that computations was really bad form :sweat_smile:

amadeuszsz commented on perception/autoware_lidar_bevfusion/src/bev_ops/bev_pool_cuda.cu:
```diff
@@ -0,0 +1,57 @@
+    out              : output features, FloatTensor[b, d, h, w, c]
+*/
+__global__ void bev_pool_kernel(
+  int b, int d, int h, int w, int n, int c, int n_intervals, const float * __restrict__ x,
```
`b` and `n` are unused, therefore please remove these arguments. I'm aware this is copied kernel from other source, but for Autoware we might tighten clang checks in near future.

knzo25 replied:
Thanks for the catch. Now I am worried a little bit and will check the equations again just in case.
In the meantime, addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/81f8be1804b50bc9848998b241a2339ca6d2f709

amadeuszsz commented on perception/autoware_lidar_bevfusion/lib/bevfusion_trt.cpp:
```diff
@@ -0,0 +1,532 @@
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <random>
```
`autoware/lidar_bevfusion/preprocess/point_type.hpp`, `cstddef`, `fstream` and `random` are unused here.

knzo25 replied:
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

amadeuszsz commented on perception/autoware_lidar_bevfusion/lib/bevfusion_trt.cpp:
```diff
@@ -0,0 +1,532 @@
+
+    nvinfer1::Dims input_indices_shape;
+    input_indices_shape.nbDims = 1;
+    input_indices_shape.d[0] = num_indices_; */
```
Please remove this comment block or mark as a part of future development.

knzo25 replied:
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/b3a7f89ec7fa46455d08e0fa501601494d67c901

amadeuszsz commented on perception/autoware_lidar_bevfusion/lib/preprocess/preprocess_kernel.cu:
```diff
@@ -0,0 +1,228 @@
+    points, {static_cast<std::int64_t>(points_size), config_.num_point_feature_size_}, tv::float32,
+    0);
+
+  auto point_limit = pc.dim(0);
```
`point_limit` is unused.

amadeuszsz commented on perception/autoware_lidar_bevfusion/lib/preprocess/preprocess_kernel.cu:
```diff
@@ -0,0 +1,228 @@
+    point_indices_data_, points_voxel_id_, vsize_xyz, grid_size, grid_stride, coors_range, true,
+    false, reinterpret_cast<std::uintptr_t>(stream_));
+
+  std::size_t real_num_voxels = static_cast<std::size_t>(std::get<0>(p2v_res).dim(0));
```
nit:
Use `auto` when casting. Applies to all casting occurrences in this file.

knzo25 replied:
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/72d49ec117122f115315d97f1697644aae57f675

knzo25 replied:
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/1fe4eae37909ab035a237f05961a73e41fdd63f3

amadeuszsz reviewed (changes requested):
BEVFusion for Autoware is coming! 🔥 
* I confirmed it works with my data! Minor bugs with edge cases.
* As we plan to deploy C+L model in future, maybe it is better to rename package to `autoware_bevfusion` and just write in docs that camera is not available yet. Actually it is already written.
* Could you please align CUDA version with Autoware (12.4) in [spconv](https://github.com/autowarefoundation/spconv_cpp) docs and also mention in spconv docs about required dependency for building (`cuda-nvrtc-dev-12-4`)?. Despite the fact we can automate libraries installation for Autoware, I bet many people will try to use this package out of Autoware. Also you might need to release version built with CUDA 12.4 and maybe include CUDA version in [release](https://github.com/autowarefoundation/spconv_cpp/releases) name (e.g. `spconv_v2.3.8_cumm_v0.5.3_cuda_12.4_amd64`.

amadeuszsz commented on perception/autoware_lidar_bevfusion/src/lidar_bevfusion_node.cpp:
```diff
@@ -0,0 +1,341 @@
+      "The size of point_cloud_range != 6: use the default parameters.");
+  }
+  if (voxel_size.size() != 3) {
+    RCLCPP_WARN_STREAM(
```
For `voxel_size` and `point_cloud_range` it should be `RCLCPP_ERROR_STREAM` and maybe throw an exception.

amadeuszsz commented on perception/autoware_lidar_bevfusion/src/lidar_bevfusion_node.cpp:
```diff
@@ -0,0 +1,341 @@
+  }
+
+  if (this->declare_parameter<bool>("build_only", false, descriptor)) {
+    RCLCPP_INFO(this->get_logger(), "TensorRT engine was built Shutting down the node.");
```
Missing "."

knzo25 replied:
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/d6bcbdf010162817844fa383b016a6d4964883bd

amadeuszsz commented on perception/autoware_lidar_bevfusion/lib/bevfusion_trt.cpp:
```diff
@@ -0,0 +1,532 @@
+    "debug/processing_time/inference_ms", stop_watch_ptr_->toc("processing/inner", true));
+
+  if (!postProcess(det_boxes3d)) {
+    RCLCPP_WARN_STREAM(
```
I think `RCLCPP_ERROR_STREAM` is more suitable here for `preProcess()`, `inference()` and `postProcess()` calls.

knzo25 replied:
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/1fe4eae37909ab035a237f05961a73e41fdd63f3

amadeuszsz commented on perception/autoware_lidar_bevfusion/lib/bevfusion_trt.cpp:
```diff
@@ -0,0 +1,532 @@
+  using autoware::cuda_utils::clear_async;
+
+  if (!is_data_layout_compatible_with_point_xyzirc(*pc_msg)) {
+    RCLCPP_WARN_STREAM(
```
`RCLCPP_ERROR_STREAM` seems more appropriate. Maybe throwing exception as well, I don't see any point to keeping node alive.

knzo25 replied:
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/1fe4eae37909ab035a237f05961a73e41fdd63f3

amadeuszsz commented on perception/autoware_lidar_bevfusion/lib/bevfusion_trt.cpp:
```diff
@@ -0,0 +1,532 @@
+  CHECK_CUDA_ERROR(cudaStreamSynchronize(stream_));
+
+  if (num_voxels < config_.min_num_voxels_) {
+    RCLCPP_WARN_STREAM(
```
This is `RCLCPP_ERROR_STREAM`

knzo25 replied:
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/1fe4eae37909ab035a237f05961a73e41fdd63f3

amadeuszsz commented on perception/autoware_lidar_bevfusion/lib/bevfusion_trt.cpp:
```diff
@@ -0,0 +1,532 @@
+    return false;
+  }
+  if (num_voxels > config_.max_num_voxels_) {
+    RCLCPP_ERROR_STREAM(
```
This is `RCLCPP_WARN_STREAM`

knzo25 replied:
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/1fe4eae37909ab035a237f05961a73e41fdd63f3

amadeuszsz commented on perception/autoware_lidar_bevfusion/lib/bevfusion_trt.cpp:
```diff
@@ -0,0 +1,532 @@
+  CHECK_CUDA_ERROR(cudaStreamSynchronize(stream_));
+
+  if (!status) {
+    RCLCPP_WARN_STREAM(
```
This is `RCLCPP_ERROR_STREAM`. Also we already log error in the top method, so maybe we can remove this log.

knzo25 replied:
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/1fe4eae37909ab035a237f05961a73e41fdd63f3

knzo25 commented on perception/autoware_lidar_bevfusion/launch/lidar_bevfusion.launch.xml:
```diff
@@ -0,0 +1,124 @@
+
+  <group if="$(var use_pointcloud_container)">
+    <load_composable_node target="$(var pointcloud_container_name)">
+      <composable_node pkg="autoware_lidar_bevfusion" plugin="autoware::lidar_bevfusion::LidarTransfusionNode" name="lidar_bevfusion">
```
Thanks. Addressed in [a816d89](https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/a816d89c08751fa73026057ba7bccd5ce40881b0)

knzo25 commented on perception/autoware_lidar_bevfusion/schema/bevfusion.schema.dummy.json:
```diff
@@ -0,0 +1,121 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Parameters for LiDAR TransFusion Node",
```
Thanks. Addressed in [e466545](https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/e466545a4f836b2e31b561f67af6b3817b13fecd)

knzo25 commented on perception/autoware_lidar_bevfusion/launch/lidar_bevfusion.launch.xml:
```diff
@@ -0,0 +1,124 @@
+  <!-- <arg name="data_path" default="$(env HOME)/autoware_data" description="packages data and artifacts directory path"/> -->
+  <arg name="data_path" default="$(find-pkg-share autoware_lidar_bevfusion)/autoware_data" description="packages data and artifacts directory path"/>
+  <arg name="model_name" default="bevfusion" description="options: `bevfusion`"/>
+  <!-- <arg name="model_path" default="$(var data_path)/lidar_bevfusion"/> -->
```
thanks. Forgot that I did this during the initial tests of the PR
[03f6344](https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/03f63443d91be40bcd0cc534fab7564a1feabafb)

knzo25 commented on perception/autoware_lidar_bevfusion/config/detection_class_remapper.param.yaml:
```diff
@@ -0,0 +1,38 @@
+  ros__parameters:
+    allow_remapping_by_area_matrix:
+      # NOTE(knzo25): We turn all vehicles into trailers if they go over 3x12 [m^2].
+      # NOTE(knzo25): We turn cars into trucks if they have an area between 2.2 x 5.5 and 3.0 * 12.0 [m^2]
```
@amadeuszsz 
Please see  centerpoint and transfusion. The note is needed really to understand the intent of the parameters.
I can remove my name if you want, but I think it is convenient to leave the comment

amadeuszsz replied:
Right! I assumed it was some sort of temporary notes

knzo25 commented on perception/autoware_lidar_bevfusion/config/bevfusion_ml_package.param.yaml:
```diff
@@ -0,0 +1,24 @@
+    num_proposals: 500
+    out_size_factor: 8
+    max_points_per_voxel: 10
+
```
My pre-commit keeps adding it. Are you sure it should not be there? (it also passes pre-commit in the CI/CD)

knzo25 commented on perception/autoware_lidar_bevfusion/include/autoware/bev_ops/bev_pool_cuda.hpp:
```diff
@@ -0,0 +1,28 @@
+
+#include <cuda_runtime_api.h>
+
+void bev_pool(
```
I would like to keep it as the source to minimize diffs. Additionally, while in autoware we are supposed to use snake case,  tensorrt uses camel case, so it will not be uniform :cry:

knzo25 commented on perception/autoware_lidar_bevfusion/include/autoware/bev_ops/bev_pool_cuda.hpp:
```diff
@@ -0,0 +1,28 @@
+void bev_pool(
+  int b, int d, int h, int w, int n, int c, int n_intervals, const float * x,
+  const int * geom_feats, const int * interval_starts, const int * interval_lengths, float * out,
+  cudaStream_t & stream);
```
Addressed in 
ac4f3a7

knzo25 commented on perception/autoware_lidar_bevfusion/include/autoware/lidar_bevfusion/postprocess/non_maximum_suppression.hpp:
```diff
@@ -0,0 +1,82 @@
+{
+using autoware_perception_msgs::msg::DetectedObject;
+
+// TODO(knzo25): Update this implementation to mach changes from centerpoint
```
Thanks. Actually, this is something that was pending and I forgot all about it. 
The centerpoint counterpart is https://github.com/autowarefoundation/autoware_universe/pull/9595

Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/6a2669c5cc271e0b678e0b92565c471438666dc6

knzo25 commented on perception/autoware_lidar_bevfusion/include/autoware/lidar_bevfusion/preprocess/precomputed_features.hpp:
```diff
@@ -0,0 +1,89 @@
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include <iostream>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/include/autoware/lidar_bevfusion/preprocess/voxel_generator.hpp:
```diff
@@ -0,0 +1,64 @@
+#include "autoware/lidar_bevfusion/bevfusion_config.hpp"
+#include "autoware/lidar_bevfusion/preprocess/pointcloud_densification.hpp"
+#include "autoware/lidar_bevfusion/preprocess/preprocess_kernel.hpp"
+#include "autoware/lidar_bevfusion/ros_utils.hpp"
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/include/autoware/tensorrt_plugins/get_indices_pairs_implicit_gemm_plugin.hpp:
```diff
@@ -0,0 +1,140 @@
+#include <NvInferRuntimePlugin.h>
+#include <cuda_runtime.h>
+
+#include <array>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/include/autoware/tensorrt_plugins/implicit_gemm_plugin.hpp:
```diff
@@ -0,0 +1,139 @@
+#include <spconvlib/spconv/csrc/sparse/convops/gemmops/GemmTunerSimple.h>  // cSpell:ignore spconvlib
+#include <spconvlib/spconv/csrc/sparse/convops/spops/ConvGemmOps.h>
+
+#include <array>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/include/autoware/tensorrt_plugins/plugin_utils.hpp:
```diff
@@ -0,0 +1,35 @@
+#include <NvInferRuntime.h>
+
+#include <cstdint>
+#include <stdexcept>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/include/autoware/tensorrt_plugins/quick_cumsum_cuda_plugin.hpp:
```diff
@@ -0,0 +1,123 @@
+#include <NvInferRuntimePlugin.h>
+#include <cuda_runtime.h>
+
+#include <array>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/src/tensorrt_plugins/get_indices_pairs_implicit_gemm_plugin_creator.cpp:
```diff
@@ -0,0 +1,320 @@
+#include <cstring>
+#include <exception>
+#include <iostream>
+#include <mutex>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/src/tensorrt_plugins/get_indices_pairs_implicit_gemm_plugin.cpp:
```diff
@@ -0,0 +1,421 @@
+#include <cstring>
+#include <exception>
+#include <functional>
+#include <iostream>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/src/tensorrt_plugins/implicit_gemm_plugin_creator.cpp:
```diff
@@ -0,0 +1,174 @@
+#include <cstring>
+#include <exception>
+#include <iostream>
+#include <mutex>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/src/tensorrt_plugins/plugin_registration.cpp:
```diff
@@ -0,0 +1,73 @@
+#include <NvInferRuntime.h>
+
+#include <cstdint>
+#include <iostream>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/src/tensorrt_plugins/quick_cumsum_cuda_plugin.cpp:
```diff
@@ -0,0 +1,248 @@
+#include <cstdlib>
+#include <cstring>
+#include <exception>
+#include <iostream>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/src/tensorrt_plugins/quick_cumsum_cuda_plugin_creator.cpp:
```diff
@@ -0,0 +1,143 @@
+#include <cstring>
+#include <exception>
+#include <iostream>
+#include <mutex>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/lib/postprocess/postprocess_kernel.cu:
```diff
@@ -0,0 +1,141 @@
+#include <thrust/host_vector.h>
+#include <thrust/sort.h>
+
+#include <cstddef>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/lib/preprocess/precomputed_features.cpp:
```diff
@@ -0,0 +1,376 @@
+#include <Eigen/Dense>
+#include <unsupported/Eigen/CXX11/Tensor>
+
+#include <algorithm>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/include/autoware/lidar_bevfusion/ros_utils.hpp:
```diff
@@ -0,0 +1,67 @@
+CHECK_FIELD(InputPointType, autoware::point_types::PointXYZIRC, intensity);
+static_assert(sizeof(InputPointType) == sizeof(autoware::point_types::PointXYZIRC));
+
+// TODO(knzo25): will move this from the pointcloud preprocessor to aut autoware_point_types after
```
Thanks for the reminder. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/d573b13960723c5e9ecad3ad060bcf37b3f830e2

knzo25 commented on perception/autoware_lidar_bevfusion/package.xml:
```diff
@@ -0,0 +1,36 @@
+  <buildtool_depend>ament_cmake_auto</buildtool_depend>
+  <buildtool_depend>autoware_cmake</buildtool_depend>
+
+  <depend>autoware_cuda_utils</depend>
```
Thanks. added in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/79bd5dac10ae9bae88e2cfb7052ebd3b6101f983

knzo25 commented on perception/autoware_lidar_bevfusion/src/lidar_bevfusion_node.cpp:
```diff
@@ -0,0 +1,341 @@
+  double lidar_stamp = rclcpp::Time(pc_msg->header.stamp).seconds();
+  camera_masks_.resize(camera_info_msgs_.size());
+  for (std::size_t i = 0; i < camera_masks_.size(); ++i) {
+    camera_masks_[i] = 1.f ? (lidar_stamp - rclcpp::Time(image_msgs_[i]->header.stamp).seconds()) <
```
:scream: 
Thanks !
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/9247dd5e70ed8ca7125cdd32214eab5621365044

knzo25 commented on perception/autoware_lidar_bevfusion/src/tensorrt_plugins/get_indices_pairs_implicit_gemm_plugin_creator.cpp:
```diff
@@ -0,0 +1,320 @@
+  plugin_attributes_.emplace_back(
+    nvinfer1::PluginField("subm", nullptr, PluginFieldType::kINT32, 1));  // cSpell:ignore subm
+  plugin_attributes_.emplace_back(
+    nvinfer1::PluginField("transpose", nullptr, PluginFieldType::kINT32, 1));
```
Thanks.
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/96115a735ac839a815fcf2ed41b2d5df3102d81f

knzo25 commented on perception/autoware_lidar_bevfusion/src/tensorrt_plugins/quick_cumsum_cuda_plugin_creator.cpp:
```diff
@@ -0,0 +1,143 @@
+  plugin_attributes_.emplace_back(
+    nvinfer1::PluginField("height", nullptr, PluginFieldType::kINT32, 1));
+  plugin_attributes_.emplace_back(
+    nvinfer1::PluginField("width", nullptr, PluginFieldType::kINT32, 1));
```
Thanks.
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/96115a735ac839a815fcf2ed41b2d5df3102d81f

knzo25 commented on perception/autoware_lidar_bevfusion/src/tensorrt_plugins/implicit_gemm_plugin.cpp:
```diff
@@ -0,0 +1,331 @@
+#include <cstdlib>
+#include <cstring>
+#include <exception>
+#include <iostream>
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/commit/c1ee01790ab25ced485c953332c8b38cc59c30fe

knzo25 commented on perception/autoware_lidar_bevfusion/include/autoware/tensorrt_plugins/get_indices_pairs_implicit_gemm_plugin.hpp:
```diff
@@ -0,0 +1,140 @@
+constexpr char const * const kGET_INDICES_PAIRS_IMPLICIT_GEMM_PLUGIN_VERSION{"1"};
+constexpr char const * const kGET_INDICES_PAIRS_IMPLICIT_GEMM_PLUGIN_NAMESPACE{""};
+
+namespace nvinfer1
```
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/9a824070cdf0999e0df2781a1cb6de2fcbac585c

knzo25 commented on perception/autoware_lidar_bevfusion/lib/postprocess/postprocess_kernel.cu:
```diff
@@ -0,0 +1,141 @@
+{
+  is_score_greater(float t) : t_(t) {}
+
+  __device__ bool operator()(const Box3D & b) { return b.score > t_; }
```
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/090cd7453cde9b9c57467110ec59b0f5462ecee2

knzo25 commented on perception/autoware_lidar_bevfusion/lib/preprocess/pointcloud_densification.cpp:
```diff
@@ -0,0 +1,111 @@
+#include <pcl_ros/transforms.hpp>
+#include <tf2_eigen/tf2_eigen.hpp>
+
+#include <boost/optional.hpp>
```
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/5ae9928a48729520b50141892cfd0b06c5964cea

knzo25 commented on perception/autoware_lidar_bevfusion/lib/preprocess/preprocess_kernel.cu:
```diff
@@ -0,0 +1,228 @@
+  const InputPointType * input_data, std::size_t points_size, float time_lag,
+  const float * transform_array, float * output_points)
+{
+  dim3 blocks(divup(points_size, config_.threads_per_block_));
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/2a6f934e9ef1529457b724637aa57da02531bfc4

knzo25 replied:
Also https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/a8d4bcb9eab2e11dbe38a58c4d83a52cacf7d053

knzo25 commented on perception/autoware_lidar_bevfusion/lib/preprocess/voxel_generator.cpp:
```diff
@@ -0,0 +1,89 @@
+    auto sweep_num_points = pc_cache_iter->num_points;
+    auto output_offset = point_counter * config_.num_point_feature_size_;
+
+    if (point_counter + sweep_num_points >= static_cast<std::size_t>(config_.cloud_capacity_)) {
```
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/4387a8ff3d1a70c27bd79bf8b3e7bd036eb43cf4

knzo25 commented on perception/autoware_lidar_bevfusion/lib/preprocess/voxel_generator.cpp:
```diff
@@ -0,0 +1,89 @@
+        rclcpp::get_logger("lidar_bevfusion"), "Exceeding cloud capacity. Used "
+                                                 << pd_ptr_->getIdx(pc_cache_iter) << " out of "
+                                                 << pd_ptr_->getCacheSize() << " sweep(s)");
+      break;
```
Thanks. Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/c473ea4cd2a9499feab0c4652691b123446c19a3

knzo25 commented:
@amadeuszsz 
Thanks for the thorough review.
I think I have addressed all the comments bar the data used to train the model. I will update that as soon as I update the latest model and make final tests

amadeuszsz commented on perception/autoware_lidar_bevfusion/lib/bevfusion_trt.cpp:
```diff
@@ -0,0 +1,508 @@
+  if (num_points == 0) {
+    RCLCPP_ERROR(
+      rclcpp::get_logger("lidar_bevfusion"),
+      "Empty sweep points (check the capacity of the buffer) . Skipping detection.");
```
```suggestion
      "Empty sweep points (check the capacity of the buffer) . Skipping detection.");
```
```suggestion
      "Empty sweep points (check the capacity of the buffer). Skipping detection.");
```

Zhao-Qihao commented:
@knzo25 Hi, what's the version of tensorrt does the bevfusion package use?

amadeuszsz commented:
> @amadeuszsz Thanks for the thorough review. I think I have addressed all the comments bar the data used to train the model. I will update that as soon as I update the latest model and make final tests

Thanks!

Could you also please share you thoughts about my review comment?
> - ~As we plan to deploy C+L model in future~ (it is already deployed), maybe it is better to rename package to autoware_bevfusion
> - Could you please align CUDA version with Autoware (12.4) in [spconv](https://github.com/autowarefoundation/spconv_cpp) docs and also mention in spconv docs about required dependency for building (cuda-nvrtc-dev-12-4)?. Despite the fact we can automate libraries installation for Autoware, I bet many people will try to use this package out of Autoware. Also you might need to release version built with CUDA 12.4 and maybe include CUDA version in [release](https://github.com/autowarefoundation/spconv_cpp/releases) name (e.g. spconv_v2.3.8_cumm_v0.5.3_cuda_12.4_amd64.

After docs update I will check PR again and approve.

knzo25 commented:
@amadeuszsz 
Addressed the final comments as follows:
 
> As we plan to deploy C+L model in future (it is already deployed), maybe it is better to rename package to autoware_bevfusion

Addressed in [2532150](https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/253215061ba8b10e2d9dc5639552cceec1e9a5dd)

> Could you please align CUDA version with Autoware (12.4) in [spconv](https://github.com/autowarefoundation/spconv_cpp) docs and also mention in spconv docs about required dependency for building (cuda-nvrtc-dev-12-4)?. Despite the fact we can automate libraries installation for Autoware, I bet many people will try to use this package out of Autoware. Also you might need to release version built with CUDA 12.4 and maybe include CUDA version in [release](https://github.com/autowarefoundation/spconv_cpp/releases) name (e.g. spconv_v2.3.8_cumm_v0.5.3_cuda_12.4_amd64.

spconv is supposedly compatible with all cuda-12 versions.


Finished the config files and the docs in [fb59f37](https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/fb59f37289787542a9cda02953cea2fb61966647) and [8db3fcf](https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/8db3fcf1fa1ce14625623a55f0c38b637366a334)

knzo25 commented on perception/autoware_lidar_bevfusion/README.md:
```diff
@@ -0,0 +1,81 @@
+
+## Trained Models
+
+TODO
```
Finished the config files and the docs in [fb59f37](https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/fb59f37289787542a9cda02953cea2fb61966647) and [8db3fcf](https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/8db3fcf1fa1ce14625623a55f0c38b637366a334)

knzo25 commented:
@Zhao-Qihao 
It requires TensorRT > 10
To check the exact requirements check the setup script of autoware

amadeuszsz commented:
@knzo25 
Thanks!

Can we get rid of dummy schema by renaming schema and config to `ml_package_bevfusion*`? Or we need to strictly follow naming convention?

Zhao-Qihao commented:
@knzo25 Could you release the new training config for the latest v2.onnx model? The onnx model I exported using [https://github.com/tier4/AWML/tree/main/projects/BEVFusion](https://github.com/tier4/AWML/tree/main/projects/BEVFusion](url) is different from the v2.onnx in autuware_bevfusion. Also, could you release the .pth file of the trained model? That way, I can continue training my own data on top of loading from your model.

Zhao-Qihao commented:
> @Zhao-Qihao The main branch of awml is not compatible with this PR Since it takes time to make public links and they become read-only once I do so, I will only make it public once the PR is about to be merged. pth will not be public on this PR

Thank you very much for your response!

knzo25 commented:
@amadeuszsz 
Addressed schema concerns in https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/cccda2fb6a9abbd845a610b38763784df02cc1ae, https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/ed31a70fc61659a558eb62a2c9e27ac7d94a68fa, and  https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/1ca7eb4680466a2ae634db6a2f2b449bec6d79a0

knzo25 commented:
@amadeuszsz 
Right !
Somehow forgot about that even though I read the schema implementation earlier today :sweat_smile: 
Addressed in [7997c93](https://github.com/autowarefoundation/autoware_universe/pull/10024/commits/7997c939f8bb2a81d7424add6248927039eb4ff3)

amadeuszsz reviewed (approved):
LGTM 🚀 

Great contribution!

Please address the remaining issues (before or after merge):
* Fix spconv installation (missing in Docker, probably for local installation too). Playbook is [not loaded](https://github.com/autowarefoundation/autoware/blob/d5e8c694ae3721adda99cba817f71841604c4620/.github/workflows/autoware-base.yaml#L7). This might require modification of `setup-dev-env.sh`. However, I think instead of triggering playbook, we can just add role to `openadkit.yaml` and `universe.yaml`. For better understanding, you can follow any occurrence of `geographiclib` in `autoware/ansible`. If I'm mistaken, I believe you will be guided during PR review in awf/autoware.
* Add artifacts to ansible task.
* Update information about artifacts in bevfusion markdown file.

Zhao-Qihao commented:
@knzo25 Hi, I found a bug when training the lidar branch using [lidar config file](https://github.com/knzo25/autoware-ml/blob/feat/bevfusion_cherrypick_updates/projects/BEVFusion/configs/nuscenes/bevfusion_lidar_voxel0075_second_secfpn_1xb1-cyclic-20e_nus-3d.py).

This might be because you tested the camera-lidar model using the v2 model, but did not test the lidar-only model. And I found that the root cause of the error is that in [bevfusion.py](https://github.com/knzo25/autoware-ml/blob/feat/bevfusion_cherrypick_updates/projects/BEVFusion/bevfusion/bevfusion.py) there should not be a ```depth_loss``` when using only the lidar branch.
![image-20250417115729035](https://github.com/user-attachments/assets/e5683a48-cdb2-43c6-ac52-5023b6267e24)

you can easily fix this bug by doing like this in [bevfusion.py](https://github.com/knzo25/autoware-ml/blob/feat/bevfusion_cherrypick_updates/projects/BEVFusion/bevfusion/bevfusion.py)
![image](https://github.com/user-attachments/assets/37369f02-7fdb-4056-a2e0-63824b69805f)

Cause it's a minor change, so I didn't make a PR in your repo [autoware-ml](https://github.com/knzo25/autoware-ml/blob/feat/bevfusion_cherrypick_updates).

By the way, could you release your BEVFusion-CL.pth file in [doc](https://github.com/knzo25/autoware-ml/blob/feat/bevfusion_cherrypick_updates/projects/BEVFusion/docs/BEVFusion-CL/v0/base.md), cause I have poor compute GPU, and it cost a lot of time to make a pretrained model.

knzo25 commented:
@Zhao-Qihao 
Thanks for the notice. I though I had fixed that before :cry: 
Regarding the pth, I do not think that we have anything against releasing pths, but this is not the place to ask for those, nor I am capable of making that decision

lhn0323 commented:
@knzo25 Hi,First of all, thank you for this PR and bringing BEVFusion to Autoware! It is a massive contribution.

I am currently testing this package on an NVIDIA Jetson Orin (32GB) using the humble branch. I trained my own model using mmdetection3d and exported bevfusion_camera_lidar_fixed.onnx and bevfusion_camera_lidar.onnx, subsequently using the fixed version.

However, I am observing significant latency issues that make it difficult to deploy for real-time applications. I would like to share my findings and ask for your advice on potential optimizations.

1. Environment & Setup

Hardware: NVIDIA Jetson Orin (32GB RAM).

Software: Autoware Universe (Humble), TensorRT 8.5+, CUDA 12.6 (JetPack 6.1).

Model Config: ResNet50 Backbone, point_cloud_range: [-54.0, -54.0, -3.0, 54.0, 54.0, 5.0] voxel_size: [0.075, 0.075, 0.2] num_cameras: 5  raw_image_height: 1536 raw_image_width: 1920 img_aug_scale_x: 0.36 img_aug_scale_y: 0.36

Export Method: Single monolithic ONNX using mmdeploy with spconv plugins enabled.

Runtime Config: trt_precision: fp16.

2. Performance Observations I benchmarked the inference time (pure GPU execution time inside detect):

LiDAR-only Mode: ~80 ms

Fusion Mode (Camera + LiDAR): ~290 ms

3. Analysis of the Bottleneck: The Fusion mode introduces an extra ~220ms latency compared to the LiDAR-only mode. This suggests that the Camera Branch (Image Backbone + View Transform) is the major bottleneck. From the logs, I confirmed that libautoware_tensorrt_plugins.so is loaded, and plugins like ImplicitGemm (for SpConv) are active.

4. Questions & Discussion: Given the hardware constraints of the Orin 32G, I have the following questions:
      
    (1) Benchmarks on Orin: Have you performed any latency benchmarks specifically on the NVIDIA Orin platform? If so, could you share the expected inference times for reference? I am trying to determine if my current result (~290ms) is within the expected range for this implementation or if there is a configuration issue.

   (2) Image branch optimization and model splitting: The overhead of the camera branch, which is approximately 220 milliseconds, seems excessively high on the Orin platform. I noticed that the latest code branch (specifically in the bevfusion_node.cpp parameters) supports loading a separate image_backbone_onnx_path. Is this design intended to decouple the image backbone from the fusion head to enable parallel execution (for example, allowing the image backbone to run in a separate CUDA stream while the lidar branch is processing)? Compared to the old branch monolithic ONNX scheme I am currently using, can switching to this model splitting architecture significantly reduce latency?

   (3) INT8 Quantization Support: To achieve a target latency of <100ms on Orin, the image backbone typically requires INT8 quantization. Does the current workflow support exporting the BEVFusion model (or at least the image backbone) with INT8 precision?

   If yes: Is there a reference configuration or script for the calibration process compatible with this node?

   If no: Is it feasible to integrate a custom INT8 image backbone (exported via other tools) into the current bevfusion_node pipeline to meet the real-time requirement?

   (4) LiDAR Branch Latency: Even the 80ms latency for the LiDAR-only branch seems slightly high compared to NVIDIA's native Lidar_AI_Solution. Is this performance gap mainly attributed to the overhead of using TensorRT plugins (Generic ImplicitGemm) for Sparse Convolutions versus NVIDIA's highly optimized, hand-written CUDA kernels (SCN)?
